File: tests/nim/operators_test.nim
# Nim 隐式双元操作符测试套件
# 覆盖所有操作符的正向、反向及边界场景

import ../../vools/bridge/nim/operators/operators
import strformat

proc repeatChar(c: char, n: int): string =
  result = newString(n)
  for i in 0..<n:
    result[i] = c

var passed = 0
var total = 0

proc test(name: string, body: proc(): bool) =
  total += 1
  if body():
    passed += 1
    echo fmt"  [PASS] {name}"
  else:
    echo fmt"  [FAIL] {name}"
    quit(1)

echo repeatChar('=', 60)
echo "Nim Implicit Operators Test Suite"
echo repeatChar('=', 60)

echo "\n--- Function-to-Function Operators ---"

test("#gt composes zero-arg to single-arg", proc(): bool =
  let genA = proc(): string = "Hello"
  let genB = proc(s: string): string = s & ", World!"
  let merged = compose01(genA, genB)
  merged() == "Hello, World!"
)

test("~gt composes single-arg functions", proc(): bool =
  let f = proc(s: string): string = s & " Are"
  let g = proc(s: string): string = s & " You?"
  let merged = compose11(f, g)
  merged("How") == "How Are You?"
)

test("~~gt composes two-arg to single-arg", proc(): bool =
  let f = proc(d, f0: string): string = d & " " & f0
  let g = proc(s: string): string = s & "!"
  let merged = compose21(f, g)
  merged("Hello", "World") == "Hello World!"
)

echo "\n--- Value-to-Function Pipe Operators ---"

test("pipe applies value to function", proc(): bool =
  let result = pipe("hello", proc(s: string): string = s.toUpper())
  result == "HELLO"
)

test("pipeMap maps seq", proc(): bool =
  let result = pipeMap(@[1, 2, 3], proc(x: int): int = x * 2)
  result == @[2, 4, 6]
)

test("pipeFilter filters seq", proc(): bool =
  let result = pipeFilter(@[1, 2, 3, 4], proc(x: int): bool = x mod 2 == 0)
  result == @[2, 4]
)

test("pipeFlat flattens and maps", proc(): bool =
  let result = pipeFlat(@[@[1, 2], @[3, 4]], proc(x: int): int = x * 10)
  result == @[10, 20, 30, 40]
)

test("pipeReduce reduces seq", proc(): bool =
  let result = pipeReduce(@[1, 2, 3, 4], proc(x, y: int): int = x + y)
  result == 10
)

test("pipeFold folds with curried folder", proc(): bool =
  proc folder(init: int): proc(acc: int, x: int): int =
    result = proc(acc: int, x: int): int = acc + x
  let sumFunc = pipeFold(@[1, 2, 3, 4], folder)
  sumFunc(0) == 10 and sumFunc(10) == 20
)

echo repeatChar('=', 60)
echo fmt"Total passed: {passed}/{total}"
echo repeatChar('=', 60)
if passed == total:
  echo "ALL TESTS PASSED"
else:
  quit(1)

File: tests/kotlin/OperatorsTest.kt
package com.example.operators

/**
 * Kotlin 隐式双元操作符测试套件
 * 覆盖所有操作符的正向、反向及边界场景
 */
fun main() {
    println("=".repeat(60))
    println("Kotlin Implicit Operators Test Suite")
    println("=".repeat(60))

    var passed = 0
    var total = 0

    fun test(name: String, block: () -> Unit) {
        total++
        try {
            block()
            passed++
            println("  [PASS] $name")
        } catch (e: Throwable) {
            println("  [FAIL] $name: ${e.message}")
            throw e
        }
    }

    println("\n--- Function-to-Function Operators ---")

    test("#gt composes zero-arg to single-arg") {
        val genA = { "Hello" }
        val genB: (String) -> String = { "$it, World!" }
        val merged = Operators.`#gt`(genA, genB)
        assert(merged() == "Hello, World!")
    }

    test("~gt composes single-arg functions") {
        val f: (String) -> String = { "$it Are" }
        val g: (String) -> String = { "$it You?" }
        val merged = Operators.`~gt`(f, g)
        assert(merged("How") == "How Are You?")
    }

    test("~~gt composes two-arg to single-arg") {
        val f: (String, String) -> String = { d, f0 -> "$d $f0" }
        val g: (String) -> String = { "$it!" }
        val merged = Operators.`~~gt`(f, g)
        assert(merged("Hello", "World") == "Hello World!")
    }

    println("\n--- Value-to-Function Pipe Operators ---")

    test("pipe applies value to function") {
        val result = Operators.pipe("hello") { it.uppercase() }
        assert(result == "HELLO")
    }

    test("pipeMap maps List") {
        val result = Operators.pipeMap(listOf(1, 2, 3)) { it * 2 }
        assert(result == listOf(2, 4, 6))
    }

    test("pipeFilter filters List") {
        val result = Operators.pipeFilter(listOf(1, 2, 3, 4)) { it % 2 == 0 }
        assert(result == listOf(2, 4))
    }

    test("pipeFlat flattens and maps") {
        val result = Operators.pipeFlat(listOf(listOf(1, 2), listOf(3, 4))) { it * 10 }
        assert(result == listOf(10, 20, 30, 40))
    }

    test("pipeReduce reduces List") {
        val result = Operators.pipeReduce(listOf(1, 2, 3, 4)) { a: Int, b: Int -> a + b }
        assert(result == 10)
    }

    test("pipeFold folds with curried folder") {
        fun folder(init: Int): (Int, Int) -> Int = { acc, x -> acc + x }
        val sumFunc = Operators.pipeFold(listOf(1, 2, 3, 4), ::folder)
        assert(sumFunc(0) == 10)
        assert(sumFunc(10) == 20)
    }

    println("=".repeat(60))
    println("Total passed: $passed/$total")
    println("=".repeat(60))
    if (passed == total) println("ALL TESTS PASSED")
    else throw RuntimeException("Some tests failed")
}
